integration_test/ESI/formats.py

```diff
@@ -0,0 +1,242 @@
+        self.srcdir = os.path.dirname(self.file)
+        self.execdir = test.getExecPath()
+        self.runs = list()
+        self.sources = list()
+        self.cppSources = list()
+        self.top = "main"
+
+    def parse(self):
+        """Parse a test file. Look for comments we recognize anywhere in the
+        file."""
+        fileReader = open(self.file, "r")
+        for line in fileReader:
+            # TOP is the top module.
+            if m := re.match(r"^//\s*TOP:(.*)$", line):
+                self.top = m.group(1)
+            # FILES are the additional RTL files (if any). If specified, must
+            # include the current file. These files are either absolute or
+            # relative to the current file.
+            if m := re.match(r"^//\s*FILES:(.*)$", line):
+                self.sources.extend(m.group(1).split())
+            # C++ driver files to feed to Verilator. Defaults to the simple
+            # driver.cpp. Same path rules as FILES.
+            if m := re.match(r"^//\s*CPP:(.*)$", line):
+                self.cppSources.extend(m.group(1).split())
+            # Run this Python line.
+            if m := re.match(r"^//\s*PY:(.*)$", line):
+                self.runs.append(m.group(1).strip())
+        fileReader.close()
+        return Test.Result(Test.PASS, "")
+
+    def compile(self):
+        """Compile the simulation with Verilator. Let Verilator do the whole
+        thing and produce a binary which 'just works'. This is sufficient for
+        simple-ish C++ drivers."""
+
+        # Assemble a list of sources (RTL and C++), applying the defaults and
+        # path rules.
+        if len(self.sources) == 0:
+            sources = [self.file]
+        else:
+            sources = [(src if os.path.isabs(src) else os.path.join(
+                self.srcdir, src)) for src in self.sources]
+        if len(self.cppSources) == 0:
+            cppSources = [os.path.join(
+                os.path.dirname(__file__), "..", "driver.cpp")]
+        else:
+            cppSources = [(src if os.path.isabs(src) else os.path.join(
+                self.srcdir, src)) for src in self.cppSources]
+
+        # Include the cosim DPI SystemVerilog files.
+        cfg = self.test.config
+        cosimInclude = os.path.join(
+            cfg.circt_include_dir, "circt", "Dialect", "ESI", "cosim")
+        sources.insert(0, os.path.join(cosimInclude, "Cosim_DpiPkg.sv"))
+        sources.insert(1, os.path.join(cosimInclude, "Cosim_Endpoint.sv"))
+
+        # Format the list of sources.
+        sources = " ".join(sources + cppSources)
+        os.makedirs(self.execdir, exist_ok=True)
+
+        # Run verilator to produce an executable. Requires a working Verilator
+        # install in the PATH.
+        vrun = subprocess.run(
+            f"{cfg.verilator_path} --cc --top-module {self.top} --build --exe {sources} {cfg.esi_cosim_path} -LDFLAGS '-Wl,-rpath={cfg.circt_shlib_dir}'".split(),
```

Comment:
i was playing around with verilator and i thought i needed the `--trace` flag here to be able to enable VCD dumps from verilator. is that necessary or was i maybe screwing something up? is there harm in adding it here?

Reply:
Running Verilator with --trace is indeed necessary to get VCD output in addition to the runtime calls in the software driver. --trace also (according to the manual) incurs a slight performance penalty.  This, however, would be a nice feature for all the Verilator tests. I've created an issue (#239) to tackle VCD debugging holistically.